### ISSUE TITLE
Fails to build on Win10, incompatible integer types

### ISSUE THREAD

**Issue by ShrykeWindgrace** [andrewthad/chronos#35]
Hi,
I stumbled upon this compilation failure:
```
src\Chronos\Internal\CTimespec.hs:53:30: error:
    * Couldn't match expected type `Int64' with actual type `Int32'
    * In the second argument of `(+)', namely `ns'
      In the first argument of `return', namely `((s * 1000000000) + ns)'
      In a stmt of a 'do' block: return ((s * 1000000000) + ns)
   |
53 |   return ((s * 1000000000) + ns)
   |                              ^^
```

Steps to reproduce: take a win10 machine (version 1703), then
```
$ stack unpack chronos && cd chronos
$ stack init
```
this inits to lts-14.0 (ghc-8.6.5); after that
```
$ stack build
```
and the compilation failure quickly follows.

How do I circumvent this issue? I can provide more info about my machine if needed.

Cheers!

**Comment by chessai** [comment]
Are you on a 32-bit or 64-bit machine?

**Comment by ShrykeWindgrace** [comment]
To quote "Control Panel\System and Security\System":
`64-bit Operating System, x64-based processor`

> Are you on a 32-bit or 64-bit machine?

**Comment by chessai** [comment]
Ah. That internal module was written to only work with Darwin/Posix. We need to write a clock_gettime that works on windows under mingw

**Comment by chessai** [comment]
I also have a feeling that on windows CTime may be a 32-bit integer, instead of 64-bit like on linux/macos. 

In ghci, on your windows machine, can you do the following:

```
>>> import Foreign.Storable
>>> import Foreign.C.Types
>>> sizeOf (undefined :: CTime)
```

If the call the sizeOf prints out 4, then im correct in my hunch

**Comment by ShrykeWindgrace** [comment]
```
Configuring GHCi with the following packages:
GHCi, version 8.6.5: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from C:\\Users\\${USERNAME}\\AppData\\Local\\Temp\\haskell-stack-ghci\\2a3bbd58\\ghci-script
Prelude> import Foreign.Storable
Prelude Foreign.Storable> import Foreign.C.Types
Prelude Foreign.Storable Foreign.C.Types> sizeOf (undefined :: CTime)
8
Prelude Foreign.Storable Foreign.C.Types> sizeOf (undefined :: CLong)
4
```
Looks like `CLong` is not what it seems to be =)

**Comment by chessai** [comment]
ok. it's possible that an introduction of a single `fromIntegral` could resolve this issue. could you try that?

**Comment by ShrykeWindgrace** [comment]
Indeed, I added a `fromIntegral` call for both terms of the sum, the compilation is ok, tests (with `stack test`) fail with 
```
ghc.exe:  | path\to\chronos-1.0.6\.stack-work\install\9330f1e8\lib\x86_64-windows-ghc-8.6.5\chronos-1.0.6-F85R03fXO934Yn1Un2ITJ6\HSchronos-1.0.6-F85R03fXO934Yn1Un2ITJ6.o: unknown symbol `clock_gettime'
ghc.exe: unable to load package `chronos-1.0.6'
doctest.exe: fd:4: hGetLine: end of file

chronos                   > Test suite doctest failed
```

**Comment by chessai** [comment]
@ShrykeWindgrace are you on mingw, or just windows? I think we should provide fixes for both...

**Comment by chessai** [comment]
i know how to fix this after googling. on both mingw and windows you have

SYSTEMTIME, which is a big struct that is basically equivalent to Chronos' `Datetime`, except with millisecond resolution instead of nanosecond. You can get the SYSTEMTIME like so

```
#include <windows.h>

SYSTEMTIME get_system_time() {
    SYSTEMTIME system;
    GetSystemTime(&system);
    return system;
}
```

**Comment by chessai** [comment]
i think the best thing to do would be a `CDatetimeSpec` datatype/struct into which we unpack the SYSTEMTIME. Then, on the haskell side, we call `datetimeToTime`. This would require very little C code. However, I do not have any windows machines except a VM which is somehow very slow. Perhaps someone else with a windows machine could implement and test this.

**Comment by ShrykeWindgrace** [comment]
> 
> 
> @ShrykeWindgrace are you on mingw, or just windows? I think we should provide fixes for both...

I ran your suggested `sizeOf` comparison both from `powershell` and `git bash`, both give the same result. I think that it comes from the `stack` - when it installs ghc, it pulls its own `mingw\x86_64-w64-mingw32`, so the the outer environment becomes irrelevant when I run ghci.

**Comment by ShrykeWindgrace** [comment]
> 
> 
> i think the best thing to do would be a `CDatetimeSpec` datatype/struct into which we unpack the SYSTEMTIME. Then, on the haskell side, we call `datetimeToTime`. This would require very little C code. However, I do not have any windows machines except a VM which is somehow very slow. Perhaps someone else with a windows machine could implement and test this.

I'll try to implement your suggestions, but my knowledge of C and of underlying windows internal machinery is rather limited. If I manage to whip a working solution, I'll let you know=)

**Comment by chessai** [comment]
I could probably write it, just couldn't compile it.

**Comment by chessai** [comment]
I asked a friend to work on this, PR coming soon

**Comment by chessai** [comment]
I forgot that the win32 library exists

**Comment by chessai** [comment]
closed in #36
